File: common/src/au/com/codeka/common/Vector2.java
package au.com.codeka.common;

import java.util.Locale;

import au.com.codeka.common.ObjectPool;
import au.com.codeka.common.ObjectPool.Pooled;

/**
 * Represents a 2-dimensional vector.
 */
public class Vector2 implements ObjectPool.Pooled {
    public static ObjectPool<Vector2> pool = new ObjectPool<Vector2>(250, new Vector2Creator());

    public double x;
    public double y;

    public Vector2() {
        x = y = 0.0;
    }
    public Vector2(double x, double y) {
        this.x = x;
        this.y = y;
    }
    public Vector2(Vector2 other) {
        this.x = other.x;
        this.y = other.y;
    }

    @Override
    public void reset() {
    }

    public Vector2 reset(double x, double y) {
        this.x = x;
        this.y = y;
        return this;
    }
    public Vector2 reset(Vector2 other) {
        this.x = other.x;
        this.y = other.y;
        return this;
    }

    /**
     * Gets the distance squared to the given other point. This is faster than
     * \c distanceTo (since no sqrt() is required) and still good enough for many
     * purposes.
     */
    public double distanceTo2(Vector2 other) {
        final double dx = other.x - x;
        final double dy = other.y - y;
        return dx*dx + dy*dy;
    }

    public double distanceTo(Vector2 other) {
        final double dx = other.x - x;
        final double dy = other.y - y;
        return Math.sqrt(dx*dx + dy*dy);
    }

    public double distanceTo2(double x, double y) {
        final double dx = x - this.x;
        final double dy = y - this.y;
        return dx*dx + dy*dy;
    }

    public double distanceTo(double x, double y) {
        final double dx = x - this.x;
        final double dy = y - this.y;
        return Math.sqrt(dx*dx + dy*dy);
    }

    public double length2() {
        return (x*x + y*y);
    }
    public double length() {
        return Math.sqrt(length2());
    }


    public void add(Vector2 other) {
        x += other.x;
        y += other.y;
    }
    public void add(float x, float y) {
        this.x += x;
        this.y += y;
    }

    public void subtract(Vector2 other) {
        x -= other.x;
        y -= other.y;
    }
    public void subtract(float x, float y) {
        this.x -= x;
        this.y -= y;
    }

    public void rotate(float radians) {
        float nx = (float)(x*Math.cos(radians) - y*Math.sin(radians));
        float ny = (float)(y*Math.cos(radians) + x*Math.sin(radians));
        x = nx;
        y = ny;
    }

    public void normalize() {
        scale(1.0 / length());
    }

    public void scale(double s) {
        x *= s;
        y *= s;
    }

    @Override
    public int hashCode() {
        // this avoids the boxing that "new Double(x).hashCode()" would require
        long lx = Double.doubleToRawLongBits(x);
        long ly = Double.doubleToRawLongBits(y);
        return (int) (lx ^ ly);
    }

    @Override
    public String toString() {
        return String.format(Locale.ENGLISH, "(%.4f, %.4f)", x, y);
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof Vector2)) {
            return false;
        }

        Vector2 ov = (Vector2) other;
        return (x == ov.x && y == ov.y);
    }

    public boolean equals(Vector2 other, double epsilon) {
        return Math.abs(other.x - x) < epsilon &&
                Math.abs(other.y - y) < epsilon;
    }

    /**
     * Find the angle between "a" and "b"
     * see: http://www.gamedev.net/topic/487576-angle-between-two-lines-clockwise/
     */
    public static float angleBetween(Vector2 a, Vector2 b) {
        return (float) Math.atan2(a.x * b.y - a.y * b.x,
                                  a.x * b.x + a.y * b.y);
    }

    public static float angleBetweenCcw(Vector2 a, Vector2 b) {
        return (float) Math.atan2(a.x * b.x + a.y * b.y,
                                  a.x * b.y - a.y * b.x);
    }

    static class Vector2Creator implements ObjectPool.PooledCreator {
        @Override
        public Pooled create() {
            return new Vector2();
        }
    }
}

File: common/src/main/kotlin/au/com/codeka/warworlds/common/Vector2.kt
package au.com.codeka.warworlds.common

import java.util.*
import kotlin.math.*

/** Represents a 2-dimensional vector. */
class Vector2 {
  var x: Double
  var y: Double

  constructor() {
    y = 0.0
    x = y
  }

  constructor(x: Double, y: Double) {
    this.x = x
    this.y = y
  }

  constructor(other: Vector2) {
    x = other.x
    y = other.y
  }

  fun reset(x: Double, y: Double): Vector2 {
    this.x = x
    this.y = y
    return this
  }

  fun reset(other: Vector2): Vector2 {
    x = other.x
    y = other.y
    return this
  }

  /**
   * Gets the distance squared to the given other point. This is faster than [distanceTo] (since
   * no sqrt() is required) and still good enough for many purposes.
   */
  fun distanceTo2(other: Vector2): Double {
    val dx = other.x - x
    val dy = other.y - y
    return dx * dx + dy * dy
  }

  fun distanceTo(other: Vector2?): Double {
    val dx = other!!.x - x
    val dy = other.y - y
    return Math.sqrt(dx * dx + dy * dy)
  }

  fun distanceTo2(x: Double, y: Double): Double {
    val dx = x - this.x
    val dy = y - this.y
    return dx * dx + dy * dy
  }

  fun distanceTo(x: Double, y: Double): Double {
    val dx = x - this.x
    val dy = y - this.y
    return Math.sqrt(dx * dx + dy * dy)
  }

  fun length2(): Double {
    return x * x + y * y
  }

  fun length(): Double {
    return sqrt(length2())
  }

  fun add(other: Vector2) {
    x += other.x
    y += other.y
  }

  fun add(x: Float, y: Float) {
    this.x += x.toDouble()
    this.y += y.toDouble()
  }

  fun subtract(other: Vector2) {
    x -= other.x
    y -= other.y
  }

  fun subtract(x: Float, y: Float) {
    this.x -= x.toDouble()
    this.y -= y.toDouble()
  }

  fun rotate(radians: Double) {
    val nx = x * cos(radians) - y * sin(radians)
    val ny = y * cos(radians) + x * sin(radians)
    x = nx
    y = ny
  }

  fun normalize() {
    scale(1.0 / length())
  }

  fun scale(s: Double) {
    x *= s
    y *= s
  }

  fun scale(sx: Double, sy: Double) {
    x *= sx
    y *= sy
  }

  override fun hashCode(): Int {
    // this avoids the boxing that "new Double(x).hashCode()" would require
    val lx = java.lang.Double.doubleToRawLongBits(x)
    val ly = java.lang.Double.doubleToRawLongBits(y)
    return (lx xor ly).toInt()
  }

  override fun toString(): String {
    return String.format(Locale.ENGLISH, "(%.4f, %.4f)", x, y)
  }

  override fun equals(other: Any?): Boolean {
    if (other !is Vector2) {
      return false
    }
    val ov = other
    return x == ov.x && y == ov.y
  }

  fun equals(other: Vector2?, epsilon: Double): Boolean {
    return abs(other!!.x - x) < epsilon && abs(other.y - y) < epsilon
  }

  companion object {
    /**
     * Find the angle between "a" and "b"
     * see: http://www.gamedev.net/topic/487576-angle-between-two-lines-clockwise/
     */
    fun angleBetween(a: Vector2, b: Vector2): Float {
      return atan2(a.x * b.y - a.y * b.x,
          a.x * b.x + a.y * b.y).toFloat()
    }

    fun angleBetweenCcw(a: Vector2, b: Vector2): Float {
      return atan2(a.x * b.x + a.y * b.y,
          a.x * b.y - a.y * b.x).toFloat()
    }
  }
}